sound/soc/sof/ipc.c

```diff
@@ -689,42 +831,54 @@ int snd_sof_ipc_set_get_comp_data(struct snd_sof_ipc *ipc,
 {
 	struct snd_sof_dev *sdev = ipc->sdev;
 	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
+	struct sof_ipc_ctrl_data_params sparams;
 	int send;
 	int err = 0;
 
+	send = sof_get_ctrl_ipc_dir(ctrl_type);
+	if (send < 0) {
+		dev_err(sdev->dev, "error: set/get ctrl type %d\n",
+			ctrl_type);
+		return send;
+	}
+
 	/* read or write firmware volume */
 	if (scontrol->readback_offset != 0) {
 		/* write/read value header via mmaped region */
-		send = sof_get_ctrl_ipc_dir(ctrl_type);
-		if (send < 0) {
-			dev_err(sdev->dev, "error: set/get ctrl type %d\n",
-				ctrl_type);
-			return send;
-		}
 		sof_set_get_mmap(sdev, scontrol, cdata, send);
 		return 0;
 	}
 
 	cdata->rhdr.hdr.cmd = SOF_IPC_GLB_COMP_MSG | ipc_cmd;
 	cdata->cmd = ctrl_cmd;
 	cdata->type = ctrl_type;
-	cdata->rhdr.hdr.size = scontrol->size;
 	cdata->comp_id = scontrol->comp_id;
-	cdata->num_elems = scontrol->num_channels;
+	cdata->msg_id = 0;
 
-	/* write value via slower IPC */
-	if (cdata->rhdr.hdr.size > SOF_IPC_MSG_MAX_SIZE) {
-		dev_err(sdev->dev, "error: set/get size %d too big comp %d\n",
-			cdata->rhdr.hdr.size, cdata->comp_id);
-		return -EINVAL;
+	/* calculate header and data size */
+	sof_get_ctrl_size_params(scontrol, cdata, &sparams);
+
+	/* send normal size ipc in one part */
+	if (cdata->rhdr.hdr.size <= SOF_IPC_MSG_MAX_SIZE) {
+		err = sof_ipc_tx_message(sdev->ipc, cdata->rhdr.hdr.cmd, cdata,
+					 cdata->rhdr.hdr.size, cdata,
+					 cdata->rhdr.hdr.size);
+
+		if (err < 0)
+			dev_err(sdev->dev, "error: set/get ctrl ipc comp %d\n",
+				cdata->comp_id);
+
+		return err;
 	}
 
-	err = sof_ipc_tx_message(sdev->ipc, cdata->rhdr.hdr.cmd, cdata,
-				 cdata->rhdr.hdr.size, cdata,
-				 cdata->rhdr.hdr.size);
+	/* data is bigger than max ipc size, chop to smaller pieces */
+	dev_dbg(sdev->dev, "large ipc size %u, control size %u\n",
+		cdata->rhdr.hdr.size, scontrol->size);
+
+	err = sof_set_get_large_ctrl_data(sdev, cdata, &sparams, send);
```

Comment:
This should only be used if the ABI level allows for large messages, otherwise you don't handle backwards compatibility.